### PR TITLE
Adds flexible pricing form

### LINKED ISSUE #573: Implement a user interface to submit requests for flexible pricing
As a Learner I want to be able to request financial aid for a program.

#### Designs and Mockups

Since Financial Aid is something that currently is DEDP-specific, we want to put much of the learner-facing UI around this in the CMS so that the DEDP team is able to setup their own content surrounding it. This also has the additional benefit of existing on a linkable page as opposed to living in the learner dashboard like MicroMasters.

A good option for this is to use [Wagtail's form builder](https://docs.wagtail.org/en/latest/reference/contrib/forms/).

#### Acceptance Criteria:

- [ ] A staff user should be able to go to the CMS and create a page to request financial aid
  - For now, financial aid will only be applicable to a program
- [ ] The learner should be able to see this page
  - If they are logged in:
    - If they have already submitted a financial aid application, instead of the form:
       - If their financial aid has been approved, display a message indicating they've been approved and what their new price is.
       - If their financial aid hasn't been approved yet, display a message indicating their financial aid application is pending and to contact support if they have any questions.
  - If they aren't logged in:
     - Instead of the form there should be a visual prompt/link to login / create an account in order to request financial aid.
- [ ] When the learner submits this form:
  - Override `def process_form_submission(self, form): ...` to **not** create a submission record and instead create a flexible pricing request object.

**Note:** The term "Financial Aid" is possibly specific to DEDP's use case. Where the CMS displays the corresponding text, that should be editable in the CMS, with a reasonable default.

#### Out of Scope

- No notifications to staff of the submitted financial aid request.

### PR DESCRIPTION
**Pull request by jkachel**
#### Pre-Flight checklist

- [X] Screenshots and design review for any changes that affect layout or styling
  - [X] Desktop screenshots
- [X] Migrations
  - [X] Migration is backwards-compatible with current production code
- [X] Testing
  - [X] Code is tested
  - [X] Changes have been manually tested

#### What are the relevant tickets?
Closes #573

#### What's this PR do?
Implements #573: Adds a Flexible Pricing Request Form to the available Wagtail forms.

Within the CMS, the form has customizable blocks for intro text and post-submit text, text for guest users, and text for  application in progress, application approved, and application denied states. Admins can customize the form fields, but the form must contain a Your Income and Income Currency field to capture the user's yearly income (similar to the form on MicroMasters). Upon submission, the form submission also creates a FlexiblePrice record, which can be used to track processing of the request. 

Applicants must be logged in to access the form - otherwise, the Guest User text is displayed and the user cannot submit an application. They must not have a prior submission. Applicants with a submission will see the appropriate text for the status of their submission (in progress, approved, skipped). 

#### How should this be manually tested?
Log in as an administrator and add Currency Exchange Rates via the Django admin. You'll need at least one (an easy one is USD since that's always 1-to-1). The Description field will be displayed in the front end if supplied, so, for completeness, at least two should be supplied, and one should not have a description.

Log in as an administrator and navigate to the CMS admin. Create a new child page under a given course that's in there - you should be given the Flexible Pricing Request Form. Supply the necessary data. You will need to create at least two form fields: one named Your Income and one named Income Currency. The "Your Income" field should be a number field, and the "Income Currency" should be a Country field. After saving the form, go back to the course page in the CMS and add a link to the form in the content (this can be in any part of the course page content that'll render a link, it doesn't have to be in a specific section). 

Then, navigate to the course page and to the form. You should see one of three things, depending on state:
1. _If you're not logged in:_ You should see whatever you put in for Guest Text, and no form.
2. _If you're logged in and haven't submitted the form:_ You should see the form.
3. _If you're logged in and have submitted the form:_ You should see one of Application In Progress, Application Approved, or Application Denied texts. 

If you've submitted an application, you should:
1. See the form submission in the Wagtail admin (under Forms)
2. See the form submission in Django Admin, in FlexiblePricing under submissions. (The form data will be in JSON format.)
3. See the FlexiblePrice record in Django Admin. Appropriate fields should be filled, and it should have an FK to the course you created the form under and to the form submission. 

#### Any background context you want to provide?
Styling on the page will likely need some help - the template reuses the resource page template. We may want to do some additional styling to form fields too, and perhaps add defaults in for the form blocks. 

The CMS support for programs is not there in the system as of yet - that seemed like a bigger task so this PR doesn't address that. The code can be easily extended to support program pages once those are in (and it's marked where to do that). 

#### Screenshots (if appropriate)
Wagtail UI:
Form
Required Form Fields

Form Display:
Basic Form
Post-Submit Thank You

Submission Status:
Post-Submit Approved
Post-Submit Denied
Post-Submit In progress

### CODE REVIEW

**Comment by pdpinch** [comment]
The submission response should be conditional on whether the request is automatically accepted, or needs review. 

If it needs review, we need to offer a link to docusign for the learner to upload documents for review.

**Comment by jkachel** [comment]
> The submission response should be conditional on whether the request is automatically accepted, or needs review.
> 
> If it needs review, we need to offer a link to docusign for the learner to upload documents for review.

@pdpinch Would the DocuSign location be specific to the learner? (In other words, does the code need to generate a location for the learner's documentation via API or can someone just put a link to a submit form into a text box in Wagtail?)

File: cms/models.py [hunk 1]
```diff
@@ -19,21 +22,63 @@
...
+        "FlexiblePricingRequestForm",
+        on_delete=models.CASCADE,
+        related_name="form_fields",
+    )
+    field_type = models.CharField(
+        verbose_name="Field Type", max_length=20, choices=CHOICES
+    )
+
+
+class FlexiblePricingFormBuilder(FormBuilder):
```
Comment (annagav) [inline cms/models.py:66]: could you add a docstring

File: cms/models.py [hunk 2]
```diff
@@ -19,21 +22,63 @@
...
+from flexiblepricing.models import (
+    CurrencyExchangeRate,
+    FlexiblePrice,
+    FlexiblePricingRequestSubmission,
+)
+
 log = logging.getLogger()
 
 
+class FormField(AbstractFormField):
```
Comment (annagav) [inline cms/models.py:53]: could you add a docstring

File: cms/models.py [hunk 3]
```diff
@@ -466,3 +514,172 @@ def get_context(self, request, *args, **kwargs):
...
+class InnerResourcePage(ResourcePage):
+    """
+    Interior page - same general format as ResourcePage, but can be nested under a ResourcePage
+    """
+
+    template = "resource_page.html"
+    parent_page_types = ["ResourcePage"]
+
+
+class FlexiblePricingRequestForm(AbstractForm):
```
Comment (annagav) [inline cms/models.py:528]: could you add a docstring here as well

File: cms/models.py [hunk 4]
```diff
@@ -466,3 +514,172 @@ def get_context(self, request, *args, **kwargs):
...
+    parent_page_types = ["cms.HomePage", "cms.ResourcePage", "cms.CoursePage"]
+
+    form_builder = FlexiblePricingFormBuilder
+    template_name = "flexiblepricing/flexible_pricing_request_form.html"
+    landing_page_template = "flexiblepricing/flexible_pricing_request_form_landing.html"
+
+    def get_submission_class(self):
+        return FlexiblePricingRequestSubmission
+
+    def get_template(self, request, *args, **kwargs):
```
Comment (annagav) [inline cms/models.py:571]: I think the template name is already all set from this line `template_name = "flexiblepricing/flexible_pricing_request_form.html"`.  It seems that the method is only useful if you want to change the template dynamically.

**Review by annagav — CHANGES_REQUESTED** [review]

**Comment by pdpinch** [comment]
> Would the DocuSign location be specific to the learner? 

No. It's a fixed URL for all users. You can see how it works on micromasters: 

https://github.com/mitodl/micromasters/blob/7bb555db606da884dece82ba2d00010fe949ce7b/static/js/components/dashboard/FinancialAidCard.js#L205

**Comment by jkachel** [comment]
https://github.com/mitodl/mitxonline/pull/553/commits/e0cdd101e48d42e507004a046a8fba54adb7f3ec adds a landing page template that can be displayed after the form is submitted that contains the DocuSign link and manual submission address. The commit doesn't include logic for displaying that particular template if the request won't be auto approved, since that'll be covered in https://github.com/mitodl/mitxonline/issues/542 (currently assigned to @asadiqbal08).

**Comment by jkachel** [comment]
@annagav resolved merge conflict

**Comment by annagav** [comment]
@jkachel the fields are not showing up for me:
Screen Shot 2022-05-19 at 4 55 16 PM

**Review by annagav — CHANGES_REQUESTED** [review]
Should we add to REDME an instruction for how to set it up if you a course admin?

**Comment by jkachel** [comment]
@annagav I can't tell from the screenshot - are you maybe seeing the status display instead of the form? If there's a submission for the user in there, it won't display the form. 

Adding a README is a good idea; I'll put one in the docs.

**Comment by annagav** [comment]
@jkachel I am getting error on form submission:
Screen Shot 2022-05-20 at 9 39 09 AM

**Comment by jkachel** [comment]
@annagav do you have an Income Currency field in your form? It's a requirement.

**Review by annagav — CHANGES_REQUESTED** [review]
After I was able to submit the form successfully, I got the "thank you for submission" message, I was thinking maybe if I refresh the page I'll get the status of my application. But then on each refresh of the page I created duplicate submissions.

**Review by annagav — APPROVED** [review]
👍

**Pull request merged**